sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -529,7 +530,12 @@ class Dataset[T] private[sql](
    * For Java API, use [[collectAsList]].
    * @since 1.6.0
    */
-  def collect(): Array[T] = rdd.collect()
+  def collect(): Array[T] = {
+    val tEnc = resolvedTEncoder
+    val input = queryExecution.analyzed.output
+    val bound = tEnc.bind(input)
+    queryExecution.toRdd.map(_.copy()).collect().map(bound.fromRow)
```

Comment:
the main difference is that we are collecting the rows and then running the encoders to do conversions here.